pandas/core/arrays/arrow/array.py

```diff
@@ -120,17 +120,24 @@ def cast_for_truediv(
             # Intentionally not using arrow_array.cast because it could be a scalar
             # value in reflected case, and safe=False only added to
             # scalar cast in pyarrow 13.
-            return pc.cast(arrow_array, pa.float64(), safe=False)
-        return arrow_array
+            # In arrow, common type between integral and float64 is float64,
+            # but integral type is safe casted to float64, to mirror python
+            # and numpy, we want an unsafe cast, so we cast both operands to
+            # to float64 before invoking arrow.
+            return pc.cast(arrow_array, pa.float64(), safe=False), pc.cast(
+                pa_object, pa.float64(), safe=False
+            )
+
+        return arrow_array, pa_object
 
     def floordiv_compat(
         left: pa.ChunkedArray | pa.Array | pa.Scalar,
         right: pa.ChunkedArray | pa.Array | pa.Scalar,
     ) -> pa.ChunkedArray:
         # Ensure int // int -> int mirroring Python/Numpy behavior
         # as pc.floor(pc.divide_checked(int, int)) -> float
-        converted_left = cast_for_truediv(left, right)
-        result = pc.floor(pc.divide(converted_left, right))
+        converted_left, converted_right = cast_for_truediv(left, right)
```

Comment:
https://github.com/pandas-dev/pandas/pull/56677 also fixes this for floordiv, IMO that is a more suitable fix, since cpython and numpy implement floordiv without going through a double, but this fix is similar to https://github.com/pandas-dev/pandas/pull/56647. This fix should still apply to truediv, since both numpy and python truediv does an unsafe cast to a double:

```
>>> a = pd.Series([18014398509481983], dtype="int64")
>>> b = pd.Series([18014398509481984], dtype="int64")
>>> a / b
0    1.0
dtype: float64
>>>
```